rig/machine_control/machine_controller.py

```diff
@@ -1657,51 +1659,120 @@ def get_p2p_routing_table(self, x, y):
+            the P2P table and which responded to :py:meth:`.get_chip_info`.
+        """
+        # A quick way of getting a list of working chips
+        p2p_tables = self.get_p2p_routing_table(x, y)
+
+        # Calculate the extent of the system
+        max_x = max(x_ for (x_, y_), r in iteritems(p2p_tables)
+                    if r != consts.P2PTableEntry.none)
+        max_y = max(y_ for (x_, y_), r in iteritems(p2p_tables)
+                    if r != consts.P2PTableEntry.none)
+
+        chips_info = {}
+
+        for (x, y), p2p_route in iteritems(p2p_tables):
+            if x <= max_x and y <= max_y:
+                if p2p_route != consts.P2PTableEntry.none:
+                    try:
+                        chips_info[(x, y)] = self.get_chip_info(x, y)
+                    except SCPError:
+                        # The chip was listed in the P2P table but is not
+                        # responding. Assume it is dead and don't include it in
+                        # the info returned.
+                        pass
+
+        return (max_x + 1, max_y + 1, chips_info)
+
+    def get_machine(self, x=0, y=0, default_num_cores=18,
+                    _all_chips_info=None):
         """Probe the machine to discover which cores and links are working.
 
         .. note::
             Links are reported as dead when the device at the other end of the
-            link does not respond to SpiNNaker nearest neighbour packets. This
-            may thus mistakenly report links attached to peripherals as dead.
+            link is not a SpiNNaker chip. Thus links attached to peripherals
+            are always marked as dead.
 
         .. note::
-            The probing process does not report how much memory is free, nor
-            how many processors are idle but rather the total available.
+            The returned object does not report how much memory is free, nor
+            how many cores are idle but rather the total available. See
+            :py:meth:`.get_machine_and_resource_constraints` or
+            :py:meth:`.get_resource_constraints` for a method which does.
 
         .. note::
             The size of the SDRAM and SysRAM heaps is assumed to be the same
             for all chips and is only checked on chip (x, y).
 
         .. note::
-            The chip (x, y) supplied is the one which will be where the search
-            for working chips begins. Selecting anything other than (0, 0), the
-            default, may be useful when debugging very broken machines.
+            The chip (x, y) supplied is the one where the search for working
+            chips begins. Selecting anything other than (0, 0), the default,
+            may be useful when debugging very broken machines.
 
         Parameters
         ----------
         default_num_cores : int
             The number of cores generally available on a SpiNNaker chip
             (including the monitor).
+        _all_chips_info : None
```

Comment:
I think this should be under `Other Parameters`, and should be documented as `None or ...`